torch/_dynamo/eval_frame.py

```diff
@@ -262,19 +264,125 @@ def innermost_fn(fn):
+    "log_file_name",
+    "verbose",
+    "verify_correctness",  # will not affect model, will raise RuntimeError
+    # (no silent change to compilation behaviour)
+    "cache_size_limit",
+    "accumulated_cache_size_limit",
+    "print_specializations",
+    "replay_record_enabled",
+    "cprofile",  # only wraps _compile, not graph
+    "repro_after",
+    "repro_level",
+    "repro_forward_only",
+    "repro_tolerance",
+    "same_two_models_use_fp64",
+    "error_on_recompile",  # safe because: will throw error
+    "report_guard_failures",
+    "report_all_guard_failures",
+    "base_dir",  # used for minifying / logging
+    "translation_validation",
+    "translation_validation_timeout",
+    "translation_validation_no_bisect",
+    "DEBUG_DIR_VAR_NAME",
+    "debug_dir_root",
+}
+
+dynamo_guarded_config_string_serialization_list = {
+    # "is_fbcode",  # function def (removed by ConfigModule visit)
+    "skipfiles_inline_module_allowlist",  # cannot pickle module (unused)
+    "constant_functions",  # PyCapsules (i.e. function ptrs)
+    "traceable_tensor_subclasses",  # Cannot pickle local object
+}
+
+# The config to restore to should dynamo compile / recompile when
+# executing from the compiled function's _TorchDynamoContext
+config_cache = threading.local()
+
+
+def _maybe_init_guarded_config_cache():
+    if hasattr(config_cache, "saved_config"):
+        return
+    else:
+        for c in ["saved_config", "saved_config_hash", "cached_recompile_hash"]:
+            setattr(config_cache, c, None)
+
+
 @contextlib.contextmanager
-def enable_dynamic(enable: Optional[bool] = None, export: bool = False):
-    if enable is None:
+def restore_compiled_dynamo_config(
+    first_ctx: bool,
+    saved_config: Dict[str, Any],
+    saved_config_hash: str,
+):
+    _maybe_init_guarded_config_cache()
+    # Set exactly once from top-level compile
+    is_top_level = False
+    try:
+        if first_ctx and config_cache.saved_config is None:
+            assert config_cache.saved_config_hash is None
+            is_top_level = True
+            config_cache.saved_config = saved_config
+            config_cache.saved_config_hash = saved_config_hash
+            log.debug(
+                "Set top-level compile config hash: %s", config_cache.saved_config_hash
+            )
+        elif config.verbose:
+            log.debug("Ignoring inner dynamo compile config and hash")  # TODO: remove?
         yield
-    elif enable:
-        # Assume everything is dynamic by default
-        with config.patch(assume_static_by_default=False):
-            yield
+    finally:
+        if is_top_level:
+            config_cache.saved_config_hash = config_cache.saved_config = None
+
+
+def get_config_and_hash(dynamic=None):
+    _maybe_init_guarded_config_cache()
+    # get current value of dynamo configs
+    dynamo_config = config.get_config_copy()
+    for k in dynamo_guarded_config_ignorelist:
+        dynamo_config.pop(k)
```

Comment:
it would sure be nice if the ignored configs were stored in a separate object somehow (I know this is annoying for BC reasons); then applying the saved cache would just be swapping out a dict, no need to do a bunch of writes and then undo the writes.

Reply:
Yeah it makes sense. We can have 2 dicts. Leave it for future refactor... https://github.com/pytorch/pytorch/issues/111150

Reply:
We could try to annotate each config.py config with `ignored` to add to ignore list.

Reply:
resolved in https://github.com/pytorch/pytorch/pull/111299